src/G2/Translation/Haskell.hs

```diff
@@ -250,22 +250,33 @@ mkCgGutsModDetailsClosures tr_con env modgutss = do
   tidys <- mapM (tidyProgram env) simplgutss
 #endif
 
-  let pairs = map (\((cg, md), mg) -> ( mkCgGutsClosure (mg_binds mg) cg
+  let pairs = map (\((cg, md), mg) -> ( mkCgGutsClosure (mg_binds mg) cg md
                                       , mkModDetailsClosure (mg_deps mg) md)) $ zip tidys simplgutss
   return pairs
 #endif
 
 -- | The core program in the CgGuts does not include local rules after tidying.
--- As such, we pass in the CoreProgram from the ModGuts
-mkCgGutsClosure :: CoreProgram -> CgGuts -> G2.CgGutsClosure
-mkCgGutsClosure bndrs cgguts =
+mkCgGutsClosure :: CoreProgram -> CgGuts -> ModDetails -> G2.CgGutsClosure
+mkCgGutsClosure binds cgguts md =
+  let
+      binds_rules = concatMap ruleInfoRules
+                  . map ruleInfo
+                  . map idInfo 
+                  . concatMap bindersOf $ binds
+  in
   G2.CgGutsClosure
     { G2.cgcc_mod_name = Just $ moduleNameString $ moduleName $ cg_module cgguts
     , G2.cgcc_binds = cg_binds cgguts
     , G2.cgcc_breaks = cg_modBreaks cgguts
     , G2.cgcc_tycons = cg_tycons cgguts
-    , G2.cgcc_rules = concatMap ruleInfoRules . map ruleInfo . map idInfo 
-                            . concatMap bindersOf $ bndrs }
+    -- Getting all rules is complicated by the facts that:
+    --    1) The core program in the CgGuts does not include local rules after tidying.
+    --    2) The ModDetails does not include all rules.
+    --    3) Names of functions in rules in the ModDetails may have been changed by bindings.
+    -- As such, we keep:
+    --    1) Rules from the ModDetails
+    --    2) Rules from the untidied CoreBinds, if a rule with the same name is not in ModDetails
```

Comment:
I am just a bit confused about what ModDetails and untidied CoreBinds do.

Reply:
In addition, what does it mean to be tidied and untidied in this scenario?

Reply:
Clarified this comment a bit.